The changes to add poi indices an use geographical parentZoneId.

IsJustaGhost opened the pull request:
The parentZoneIds change will only effect a select few zones .However, this can be reverted if need be.

IsJustaGhost committed:
Update LibZone.lua

Added pin indices for zones that have pins on parent map.
Updated parentZoneIds to reflect geographical location.

This update is based off of the current version on esoui

IsJustaGhost committed:
Update LibZone.lua

Updated changes to match current pts.

IsJustaGhost committed:
Update LibZone.lua

Fixed missing predefined local poiDataTable

IsJustaGhost committed:
Update LibZone.lua

Added missing buildPoiDataTable(). Another thing I missed after changing from current release to pts.

IsJustaGhost committed:
Update LibZone.lua

Lastly, clear the poiDataTable after zoneData is created.

Baertram commented on LibZone/LibZone.lua:
```diff
@@ -117,31 +118,188 @@ local function checkOtherLanguagesZoneDataAndTransferFromSavedVariables()
     end
 end
 
+-- This table contains parentZoneId modifiers for select zones.
```
Please move such data to the appropriate file: LibZone_Data.lua and do not keep it somwhere hidden between code in he main library file.

IsJustaGhost replied:
Moved

Baertram commented on LibZone/LibZone.lua:
```diff
@@ -117,31 +118,188 @@ local function checkOtherLanguagesZoneDataAndTransferFromSavedVariables()
+local wayShrineString = GetString(SI_DEATH_PROMPT_WAYSHRINE)
+local function buildPoiDataTable()
+	-- The initial data is for zones that do not match up properly by name.
+	poiDataTable = {
```
Same as above:
Please move such data to the appropriate file: LibZone_Data.lua and do not keep it somwhere hidden between code in he main library file.

IsJustaGhost replied:
Moved

Baertram commented on LibZone/LibZone.lua:
```diff
@@ -117,31 +118,188 @@ local function checkOtherLanguagesZoneDataAndTransferFromSavedVariables()
+		},
+	}
+
+   --d("[LibZone]GetAllZoneDataById, reBuildNew: " ..tostring(reBuildNew) .. ", doReloadUI: " ..tostring(doReloadUI) .. ", lang: " .. tostring(lang))
```
Remove or change name to current function's name please

IsJustaGhost replied:
Removed

Baertram commented on LibZone/LibZone.lua:
```diff
@@ -117,31 +118,188 @@ local function checkOtherLanguagesZoneDataAndTransferFromSavedVariables()
+        lib.maxZoneIndices, lib.maxZoneIds = getMaxZoneIndicesAndIds()
+    end
+    local maxZoneIndices = lib.maxZoneIndices
+    assert(maxZoneIndices ~= nil, "[\'" .. libraryName .. "\':GetAllZoneDataById]Error: Missing maxZoneIndices!")
```
Change name to current function's name please

IsJustaGhost replied:
Removed

Baertram commented on LibZone/LibZone.lua:
```diff
@@ -117,31 +118,188 @@ local function checkOtherLanguagesZoneDataAndTransferFromSavedVariables()
+	}
+
+   --d("[LibZone]GetAllZoneDataById, reBuildNew: " ..tostring(reBuildNew) .. ", doReloadUI: " ..tostring(doReloadUI) .. ", lang: " .. tostring(lang))
+    --Maximum of ZoneIds to check
```
Why is this needed again in here? It's done already before at lib:GetAllZoneDataById isn't it?
Do you need to update it here? Else just re-use lib.maxZoneIndices

IsJustaGhost replied:
Removed

Baertram commented on LibZone/LibZone.lua:
```diff
@@ -117,31 +118,188 @@ local function checkOtherLanguagesZoneDataAndTransferFromSavedVariables()
+		if type(poiName) == 'string' then
+			local score = comparePoiNames(poiName, zoneName)
+
+			if score then
```
```
if not match or (score ~= nil and score < match) then
   match = score
   result = pinInfo
end
```

IsJustaGhost replied:
Oh, that will work. If score is nil then match will remain nil.

Baertram commented on LibZone/LibZone.lua:
```diff
@@ -117,31 +118,188 @@ local function checkOtherLanguagesZoneDataAndTransferFromSavedVariables()
+--		["poiZoneIndex"] = number,
+--->}
+local function getPinInfo(zoneId, parentZoneId)
+	parentZoneId = geographicalParentZoneId[zoneId] or parentZoneId
```
Please add a nil check for zoneId and parentZoneId at the beginning so that missing ids at the call do not raise an error.
Or add an assert so an error message throws and shows.

IsJustaGhost replied:
I'll add that.
if zoneId and zoneIndexOfZoneId and zoneIndexOfZoneId ~= 1 then
-- is called prior. 
parentZoneId will be >= 0 as argument.

Baertram commented on LibZone/LibZone.lua:
```diff
@@ -183,6 +342,10 @@ function lib:GetAllZoneDataById(reBuildNew, doReloadUI)
     local addedAtLeastOne = false
+
+	-- Create the table used to add poiInfo to zoneData
+	buildPoiDataTable()
```
Is this needed to be re-done at each event_add_on_loaded, or event_player_activated -> The later would fire for a zone/map change with loading screen.
Did you test the performance with your changes and it is noticable because it scans all kinds of map's pins etc.?
Are the pins that get scanned filterable so custom pins by addons are skipped or isn't that needed?

Or is the buildPOIDataTable something one could do once offline and add to the preloaded data tables at LibZone_data.lua so it does not need to compute everything  new?

IsJustaGhost replied:
I have it nil out the table since it is only needed during build. 
I did not know it was called so often. 
That can be change. can build on load and not nil it.

Baertram replied:
What I meant is: This function lib:GetAllZoneDataById here is called for each event_add_on_loaded which will be at login, reloadui, zone change with loading screen afaik.
If the data you have added can be stored fixed at the LibZone_Data.lua tables move it there. If it needs to be updated with ingame API during event_Add_on_loaded then leave it here. Depends on what time it takes to scan that and if there will be a noticable lag then.

If your addition though needs to be run on EACH zone change you need to move the code to an event_player_activated! As this function here is not called at each zone change, only once after loading screen/login/reloadui.

lib:GetAllZoneDataById can only be called manually by addons but shouldn't to be fair.

IsJustaGhost replied:
I thought it was only ran OnLibraryLoaded unless specifically called by an addon. 
Really, the only time the table needs to exist is when updating data in the sv. It's likely close to being the same size of the SV file prior to the change. I figured that if it's not needed then dump it. The table itself does not take long to build. I've had 2 or 3 of them running on load from different sources. 
It's something worth testing further.

Baertram commented on LibZone/LibZone.lua:
```diff
@@ -220,13 +383,18 @@ function lib:GetAllZoneDataById(reBuildNew, doReloadUI)
-                        zoneDataForId.parentZone = zoneParentIdOfZoneId
-                    end
-                end
+				if zoneDataForId.parentZone == nil then
```
I'm not sure this will break other addons if we start to insert "custom parent zoneIds" here, even though they "geographically" belong to the zoneId?
e.g. if we return a parentZone 12345 now where else it was 213142342342 (aurbis) and other addons checked that somehow if parentZoneId > 600000 then ... -> we might be breaking their code.

I'd rather have this API function untouched at that place AND provide these "custom geograpically" parentZoneIds in a seperate table! Functions returning the parentZoneId should get a new boolean parameter and if you set this to true it will return the custom geographically parentZoneIds as parentZoneId, else they will the same as today.
I want this to be a forced decision of addon authors, no automatic one.

Or "is this really safe" by any ways?

IsJustaGhost replied:
This can be ignored and the poiZoneIndex can be use for correct zone.

Do the zoneIds go up that high using the zoneIndex loop?
    for zoneIndexOfZoneId=0, maxZoneIndices do

The parentZoneId is acquired by
one of geographicalParentZoneId
or if has poiInfo it is GetZoneId(zoneIndex) zoneIndex being the zoneIndexOfZoneId in buildPoiDataTable
else GetParentZoneId

IsJustaGhost replied:
In my current LibZone sv 1345 is the highest zoneId.
My original thought was that if an addon needed to, for example, check for summerset from Artaeum then they can GetParentZoneId(zoneId)
I'm going to put more thought into the parent.

Baertram replied:
Looked good so far, but I did not understand all yet :-) Thanks for your contribution!
Currently too busy with work and life to do anything with ESO and addons...

One more hint: If you have moved the data to the Libone_Data.lua files add a local pointing to it within LibZone.lua (e.g. local wayshrineName = lib.WayshrineName so that the usage is not always checking the whole lib table for "wayshrineName" on each loop call. Same for the preloaded data, if possible.

As DanBatson said as an advice:
```
my general advice for locals/globals is that it's better to make a local of the global if you plan on indexing the global more than once, because global lookup is one of the more expesnive lookups. Expesnsive in this case is still mostly consequential, so you'll never likely notice a difference. But I try to minimize repeat global lookups of the same variable when I can. As for further indexing something inside the table, my general rule of thumb is if I'm going to index that same value more than 3 times, I'll make a local for it. so rather than do:

if obj.val and objv.val < 100 and obj.val > 10 then
I will do

local val = obj.val
if val and val < 100 and val > 10 then
```

IsJustaGhost committed:
Update LibZone_Data.lua

Added lib.wayshrineString
Added lib.geographicalParentZoneIds

IsJustaGhost committed:
Update LibZone.lua

Moved lib.wayshrineString to LibZone_Data.lua
Moved lib.geographicalParentZoneIds to LibZone_Data.lua

IsJustaGhost committed:
Update LibZone.lua

Moved lib.wayshrineString to LibZone_Data.lua
Moved lib.geographicalParentZoneIds to LibZone_Data.lua

Restored parentZone
Added parentMapZoneId
- Since the purpose of this parent is to use with the map in one way or another
